### PR TITLE
feat: Migrate @strapi/admin server to typescript

### PR DESCRIPTION
**Pull request by Marc-Roig**
### What does it do?

Migrates the strapi admin package (server side) to typescript.

### CODE REVIEW

**Comment by Marc-Roig** [comment]
Some comments about this work.

- @remidej (and in general everyone working in the admin package). We will try to merge this asap so we don't have conflics, but what features are you working at the moment that impact strapi/admin ? We might have to sync on that

- At the moment, this is only the barebones to start working on the migration, and only works if you run strapi on ce (yarn develop:ce).

File: packages/core/admin/package.json
Comment (Marc-Roig) [inline packages/core/admin/package.json:60]: @innerdvations @Convly  eslint is asking me to put @strapi/types into `dependencies` and not `devDependencies`. Is that intended?

File: packages/core/admin/server/src/domain/condition/index.js
Comment (remidej) [inline packages/core/admin/server/src/domain/condition/index.js:22]: What's the reason for this change? will it be problematic to have jsdoc types during the migration?

File: packages/core/admin/server/src/utils/index.js
Comment (remidej) [inline packages/core/admin/server/src/utils/index.js:1]: Why is this needed?

**Comment by remidej** [comment]
@Marc-Roig we're currently working on the Blocks editor. So we're adding a lot of code in `admin/src/content-manager/components/BlocksEditor`, but not elsewhere. Looks like the only conflict will be to move our code into the src directory, but that shouldn't be problematic

**Comment by Marc-Roig** [comment]
@remidej perfect, and this is only the backend migration at the moment. Thank you!

File: packages/core/admin/server/src/utils/index.js
Comment (Marc-Roig) [reply packages/core/admin/server/src/utils/index.js:1]: @innerdvations wanted to ask about this.
@strpai/types injects the global strapi variable type definition, and as this file is using the strapi global, the only way I could find was to import strapi types like this

File: packages/core/admin/server/src/domain/condition/index.js
Comment (Marc-Roig) [reply packages/core/admin/server/src/domain/condition/index.js:22]: I think not, for some reason the Condition type definition was colliding with another internal Typescript definition, and was getting a BLOCK-SCOPED error.

File: packages/core/admin/server/src/utils/index.js
Comment (Convly) [reply packages/core/admin/server/src/utils/index.js:1]: You could always do `import '@strapi/types'` instead.

But ideally, we should finally just find a way to remove the strapi global from everywhere we can in V5

File: packages/core/admin/package.json
Comment (Convly) [reply packages/core/admin/package.json:60]: That's weird, is that a warning or an error?

Comment (Marc-Roig) [reply packages/core/admin/package.json:60]: eslint error

**Comment by jhoward1994** [comment]
The TS conversion is now complete for `admin/server` and `admin/ee` 🚀

EE needs more specific typing which we will handle separately after this is merged.

Would someone be able to review this? It's a lot of files but I think the most important things to check are the new types in `packages/core/admin/shared` and a general QA of the app

@remidej @innerdvations

File: packages/core/admin/shared/contracts/admin.ts
Comment (joshuaellis) [inline packages/core/admin/shared/contracts/admin.ts:29]: I think you always send an object or Logo for these, same with `project-settings`

File: packages/core/admin/ee/server/src/routes/utils.ts
Comment (joshuaellis) [inline packages/core/admin/ee/server/src/routes/utils.ts:3]: same middleware thing here

File: packages/core/admin/shared/contracts/admin.ts
Comment (joshuaellis) [inline packages/core/admin/shared/contracts/admin.ts:2]: I actually think your ValidationError for yup is part of `errors` already

File: packages/core/admin/shared/contracts/admin.ts
Comment (joshuaellis) [inline packages/core/admin/shared/contracts/admin.ts:46]: Shouldn't this actually be `Logo`?

File: packages/core/admin/shared/contracts/webhooks.ts
Comment (joshuaellis) [inline packages/core/admin/shared/contracts/webhooks.ts:6]: Can we add the URLS to make them consistent?

**Review by innerdvations — CHANGES_REQUESTED** [review]
Some general comments:

- can you try to use 'unknown' instead of 'any' wherever possible? At least that way we don't disable the type checking altogether

- similar to that, would be nice to add more 'easy' types like Koa Context and Next instead of `any` in a lot of places

- I thought we had a lint rule on @ts-expect-error requiring an explanation after, but it doesn't appear to be working here. Can you add comments for why each expect error is needed?

File: packages/core/admin/ee/server/src/bootstrap.ts
Comment (jhoward1994) [reply packages/core/admin/ee/server/src/bootstrap.ts:1]: In most other places in packages/admin we import it as `EE` so to me it makes sense to keep them all consistent

File: packages/core/admin/server/src/domain/action/provider.ts
Comment (jhoward1994) [reply packages/core/admin/server/src/domain/action/provider.ts]: This previously wasn't handled at all see - https://github.com/strapi/strapi/blob/5c5ae949bfec6a57131caacae14a36e8c1465080/packages/core/admin/server/domain/action/provider.js#L54

wdyt of the error added in https://github.com/strapi/strapi/pull/18232/commits/ecbd3670613f45c63f5bf4de5563371b6b06438d

File: packages/core/admin/shared/contracts/api-token.ts
Comment (Marc-Roig) [reply packages/core/admin/shared/contracts/api-token.ts:5]: Most of the types were copied from the previous jsdocs and we were unsure on using different types, I would appreciate if you could clean it up in another moment for sure :pepe-love:

File: packages/core/admin/ee/server/src/bootstrap.ts
Comment (Marc-Roig) [reply packages/core/admin/ee/server/src/bootstrap.ts:1]: Same, in some cases when looking for pleces where we use this we search for EE. in the codebase, it makes it easy to navigate

**Comment by jhoward1994** [comment]
Hey @Marc-Roig & @innerdvations 

This is ready to review again once the checks have passed. I've addressed the feedback in the commits since the last comment 👍🏻

**Comment by Marc-Roig** [comment]
@jhoward1994 PR looks good to me, I can't approve it as I originally created the PR , but here you have my check ✅

**Review by jhoward1994 — APPROVED** [review]

**Comment by jhoward1994** [comment]
I have done a QA of the admin package in the UI and checking the admin package routes are working.

imo we can merge this and take a similar approach to https://github.com/strapi/strapi/pull/18669#issuecomment-1812302248 regarding the FE tests

wdyt @innerdvations @Marc-Roig ?

**Review by innerdvations — APPROVED** [review]
LGTM. If you're doing another pass, we could remove msot of the jsdoc comments, but I don't think they hurt so not a big deal.

**Comment by joshuaellis** [comment]
@jhoward1994 when this passes (eventually 🤪) let's ensure we squash merge because we don't want 133 commits for this lol

**Pull request merged**
